src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs

```diff
@@ -841,58 +841,96 @@ private static bool IsConditionalCastSafeToRemove(
+
+            if (IsNullOrErrorType(originalCastExpressionTypeInfo) ||
+                IsNullOrErrorType(originalConditionalTypeInfo) ||
+                IsNullOrErrorType(rewrittenConditionalTypeInfo))
             {
                 return false;
             }
-        }
 
-        var originalCastExpressionTypeInfo = originalSemanticModel.GetTypeInfo(castExpression, cancellationToken);
-        var originalConditionalTypeInfo = originalSemanticModel.GetTypeInfo(originalConditionalExpression, cancellationToken);
-        var rewrittenConditionalTypeInfo = rewrittenSemanticModel.GetTypeInfo(rewrittenConditionalExpression, cancellationToken);
+            // when we have    a ? (T)b : c
+            // 
+            // then we want the type of the written conditional to be T as well.  And we want the final converted
+            // type of `a ? b : c` to be the same as what `a ? (T)b : c` is converted to.
+
+            if (!originalConditionalTypeInfo.ConvertedType!.Equals(rewrittenConditionalTypeInfo.ConvertedType, SymbolEqualityComparer.IncludeNullability))
+                return false;
+
+            var castType = originalSemanticModel.GetTypeInfo(castExpression, cancellationToken).Type;
+            if (IsNullOrErrorType(castType))
+                return false;
+
+            if (rewrittenSemanticModel.GetOperation(rewrittenConditionalExpression, cancellationToken) is not IConditionalOperation rewrittenConditionalOperation)
+                return false;
+
+            if (castType.Equals(rewrittenConditionalOperation.Type, SymbolEqualityComparer.IncludeNullability))
+                return true;
+
+            if (rewrittenConditionalOperation.Parent is IConversionOperation conditionalParentConversion &&
+                conditionalParentConversion.GetConversion().IsImplicit &&
+                castType.Equals(conditionalParentConversion.Type, SymbolEqualityComparer.IncludeNullability))
+            {
+                return true;
+            }
 
-        if (IsNullOrErrorType(originalCastExpressionTypeInfo) ||
-            IsNullOrErrorType(originalConditionalTypeInfo) ||
-            IsNullOrErrorType(rewrittenConditionalTypeInfo))
-        {
             return false;
         }
 
-        // when we have    a ? (T)b : c
-        // 
-        // then we want the type of the written conditional to be T as well.  And we want the final converted
-        // type of `a ? b : c` to be the same as what `a ? (T)b : c` is converted to.
+        // Returns true if we have `x ? (T)y : z` and (T) can be removed because the 'y' type is hte same as the 'z' type, and
+        // both are converted to 'T' outside of the conditional.
+        bool IsConditionalCastSafeToRemoveDueToConversionToOtherBranch()
+        {
+            // Always keep a cast of 'default'.  This can end up taking on incorrect values if it uses the type of hte other branch
```

Comment:
```suggestion
            // Always keep a cast of 'default'.  This can end up taking on incorrect values if it uses the type of the other branch
```